### PR TITLE
Add binary literals

### PR DESCRIPTION
**Pull request by NthPortal**

### CODE REVIEW

**Comment by NthPortal** [comment]
see also: scala/docs.scala-lang#1464

File: test/files/neg/literals.scala [hunk 1]
```diff
@@ -3,35 +3,39 @@ trait RejectedLiterals {
 
   def missingHex: Int    = { 0x }        // line 4: was: not reported, taken as zero
 
-  def leadingZeros: Int  = { 01 }        // line 6: no leading zero
+  def missingBin: Int    = { 0b }        // line 6: missing integer number
```
Comment (som-snytt) [inline test/files/neg/literals.scala]: I'd encourage appending to neg tests, especially to minimize the diff on the check file.

In this case, the additional manual comments in the test make this even more desirable.

I was going to try a "worksheet" style test that would update inline before/after comments, but would before be the starr or just before this commit?

File: test/files/run/literals.scala [hunk 2]
```diff
@@ -63,6 +63,25 @@ object Test {
...
+    check_success("0b_0110 == 0x6", 0b_0110, 0x6)
+    check_success("0b_0111 == 0x7", 0b_0111, 0x7)
+    check_success("0b_1000 == 0x8", 0b_1000, 0x8)
+    check_success("0b_1001 == 0x9", 0b_1001, 0x9)
+    check_success("0b_1010 == 0xa", 0b_1010, 0xa)
+    check_success("0b_1011 == 0xb", 0b_1011, 0xb)
+    check_success("0b_1100 == 0xc", 0b_1100, 0xc)
+    check_success("0b_1101 == 0xd", 0b_1101, 0xd)
+    check_success("0b_1110 == 0xe", 0b_1110, 0xe)
+    check_success("0b_1111 == 0xf", 0b_1111, 0xf)
```
Comment (som-snytt) [inline test/files/run/literals.scala:81]: What, no `0b_1111_1111`?

**Comment by som-snytt** [comment]
Here is a branch https://github.com/som-snytt/scala/tree/review/binary-literals and commit https://github.com/som-snytt/scala/commit/d385d955b012aafd4c848a3d6fb01a92c41622bd showing the small delta atop my other PR.

**Review by som-snytt — COMMENTED** [review]
This will look even better rebased on my other clean-up PR in scanners.

**Comment by som-snytt** [comment]
I also checked
```
scala> 0b1000100010001000_1000100010001000_1
       ^
       error: integer number too large

scala> 0b1000100010001000_1000100010001000_1000100010001000_1000100010001000L
res0: Long = -8608480567731124088

scala> 0b1000100010001000_1000100010001000_1000100010001000_1000100010001000_0L
       ^
       error: integer number too large
```
Maybe I'll take a few minutes this week to understand the overflow check.

I was curious if there's an internal limit, but it's commented out:
```
    protected def putChar(c: Char): Unit = {
//      assert(cbuf.size < 10000, cbuf)
      cbuf.append(c)
    }
```
where cbuf is a `StringBuilder` not an array.

File: test/files/run/literals.scala (hunk 2 above)
Comment (NthPortal) [reply test/files/run/literals.scala:81]: that's uhh... 16 times as long >.>

File: test/files/neg/literals.scala (hunk 1 above)
Comment (NthPortal) [reply test/files/neg/literals.scala]: I can do that. I guess I wanted to maintain organization, but it's probably not worth it considering the diff

File: test/files/run/literals.scala (hunk 2 above)
Comment (NthPortal) [reply test/files/run/literals.scala:81]: what I actually should have tested is `0b1 // Kenobi`

**Comment by dwijnand** [comment]
Hmm, looks like this hasn't got much interest.  I sent a bump in https://contributors.scala-lang.org/t/pre-sip-support-binary-integer-literals/3559/5 to see if anyone in the SIP committee wants to champion it, otherwise it should probably be closed as a Pre-SIP that didn't rally a champion.

**Comment by dwijnand** [comment]
Published: https://docs.scala-lang.org/sips/binary-integer-literals.html

**Comment by NthPortal** [comment]
Thanks for publishing this Dale!

**Comment by SethTisue** [comment]
Merging this would obligate the initial Scala 3.0 release to also support it, and Martin was adamant that the 3.0 plate is already full (reference: https://github.com/scala/docs.scala-lang/pull/1639).

There are starting to be more and more situations where a change needs to land first in Dotty, _then_ be considered for a possible backport to Scala 2, and it seems this is one of them.

On a positive note, the linked minutes also mention that a broader Dotty proposal around numeric literals more generally might open the door for this.

**Comment by NthPortal** [comment]
I might open a PR against Dotty if I get the time; if that happens, I will link it here.

**Comment by som-snytt** [comment]
Scala is the only place where binaries must agitate for inclusion.

**Comment by SethTisue** [comment]
for PRs that are unable to progress in 2.13.x, we are closing them but leaving them on the 2.14.0-M1 milestone so we can find them later

for context on this, see scala/scala-dev#661

**Pull request closed without merging**